File: app/src/main/java/com/wbrawner/simplemarkdown/view/fragment/SettingsFragment.java
package com.wbrawner.simplemarkdown.view.fragment;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;

import com.wbrawner.simplemarkdown.BuildConfig;
import com.wbrawner.simplemarkdown.R;

public class SettingsFragment extends PreferenceFragment
        implements SharedPreferences.OnSharedPreferenceChangeListener {
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addPreferencesFromResource(R.xml.pref_general);
        SharedPreferences sharedPreferences =
                PreferenceManager.getDefaultSharedPreferences(getActivity());
        sharedPreferences.registerOnSharedPreferenceChangeListener(this);
        setListPreferenceSummary(
                sharedPreferences,
                findPreference(getString(R.string.key_default_view))
        );
        if (!BuildConfig.ENABLE_CUSTOM_CSS) {
            getPreferenceScreen().removePreference(findPreference(getString(R.string.pref_custom_css)));
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        return inflater.inflate(R.layout.preference_list_fragment_safe, container, false);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Preference preference = findPreference(key);

        if (preference instanceof ListPreference) {
            setListPreferenceSummary(sharedPreferences, preference);
        }
    }

    private void setListPreferenceSummary(SharedPreferences sharedPreferences, Preference preference) {
        ListPreference listPreference = (ListPreference) preference;
        String storedValue = sharedPreferences.getString(preference.getKey(), "");
        if (storedValue.isEmpty()) {
            return;
        }
        int index = 0;
        try {
            index = Integer.valueOf(storedValue);
        } catch (NumberFormatException e) {
            // TODO: Report this?
            Log.e("SimpleMarkdown", "Unable to parse " + storedValue + " to integer");
        }
        String summary = listPreference.getEntries()[index].toString();
        preference.setSummary(summary);
    }
}

File: app/src/main/java/com/wbrawner/simplemarkdown/view/fragment/SettingsFragment.kt
package com.wbrawner.simplemarkdown.view.fragment

import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import androidx.appcompat.app.AppCompatDelegate
import androidx.lifecycle.lifecycleScope
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.wbrawner.simplemarkdown.BuildConfig
import com.wbrawner.simplemarkdown.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class SettingsFragment
    : PreferenceFragmentCompat(),
        SharedPreferences.OnSharedPreferenceChangeListener {
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.pref_general)
        if (BuildConfig.DEBUG) {
            preferenceScreen.addPreference(Preference(preferenceScreen.context).apply {
                title = "Force crash"
                onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    throw RuntimeException("Forced crash from settings")
                }
            })
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        lifecycleScope.launch(context = Dispatchers.IO) {
            val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(requireContext())
            sharedPreferences.registerOnSharedPreferenceChangeListener(this@SettingsFragment)
            (findPreference(getString(R.string.pref_key_dark_mode)) as? ListPreference)?.let {
                setListPreferenceSummary(sharedPreferences, it)
            }
            @Suppress("ConstantConditionIf")
            if (!BuildConfig.ENABLE_CUSTOM_CSS) {
                findPreference<Preference>(getString(R.string.pref_custom_css))?.let {
                    preferenceScreen.removePreference(it)
                }
            }
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (!isAdded) return
        val preference = findPreference(key) as? ListPreference ?: return
        setListPreferenceSummary(sharedPreferences, preference)
        if (preference.key != getString(R.string.pref_key_dark_mode)) {
            return
        }
        var darkMode: Int = if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
            AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY
        } else {
            AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
        }
        val darkModeValue = sharedPreferences.getString(preference.key, null)
        if (darkModeValue != null && darkModeValue.isNotEmpty()) {
            if (darkModeValue.equals(getString(R.string.pref_value_light), ignoreCase = true)) {
                darkMode = AppCompatDelegate.MODE_NIGHT_NO
            } else if (darkModeValue.equals(getString(R.string.pref_value_dark), ignoreCase = true)) {
                darkMode = AppCompatDelegate.MODE_NIGHT_YES
            }
        }
        AppCompatDelegate.setDefaultNightMode(darkMode)
        activity?.recreate()
    }

    private fun setListPreferenceSummary(sharedPreferences: SharedPreferences, preference: ListPreference) {
        val storedValue = sharedPreferences.getString(
                preference.key,
                null
        ) ?: return
        val index = preference.findIndexOfValue(storedValue)
        if (index < 0) return
        preference.summary = preference.entries[index].toString()
    }
}
